.github/workflows/build_for_pypi.yml

```
@@ -1,47 +1,84 @@
+---
 name: Build and Optionally Publish to PyPi
 
 on:
-  workflow_call:
-    inputs:
-      publish:
-        type: boolean
-        default: false
-        description: "Publish to PyPi"
+  pull_request:
+    branches:
+      - main
+  push:
+    # Pattern matched against refs/tags
+    tags:
+      - 'v[0-9]+.[0-9]+.[0-9]+'
 
 jobs:
-  build_for_pypi:
+  build_src_for_pypi:
     runs-on: ubuntu-latest
-    permissions:
-      id-token: write # This is required for requesting the JWT
-      contents: read  # This is required for actions/checkout
     steps:
       - uses: actions/checkout@v4
         with:
+          persist-credentials: false
           submodules: true
-      - name: Set up Python 3.11
-        uses: actions/setup-python@v4
-        with:
-          python-version: "3.11"
       - name: Install dependencies
         run: |
-          pip install -r requirements.txt
-          python setup.py build
-          python setup.py develop
-      - name: Build distributions for different platforms
+          pip install build
+      - name: Build src dist
         run: |
-          pip install wheel
-          python setup.py sdist bdist_wheel --plat-name=manylinux2014_x86_64
-          python setup.py bdist_wheel --plat-name=macosx-12.6-universal2
-          python setup.py bdist_wheel --plat-name=win_amd64
-      - name: Move src file to match PEP625
+          python -m build --sdist
+        env:
+          PIP_CONSTRAINT: requirements.txt
+      - name: Store the distribution packages
+        uses: actions/upload-artifact@v4
+        with:
+          name: cibw-sdist
+          path: ./**/*.tar.gz
+
+  build_dist_for_pypi:
+    needs:
+      - build_src_for_pypi
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os:
+          - macos-13
+          - macos-latest
+          - ubuntu-24.04-arm
+          - ubuntu-latest
+          - windows-latest
+    steps:
+      - name: Download the sdist
+        uses: actions/download-artifact@v4
+        with:
+          name: cibw-sdist
+      - name: Get sdist filename
+        id: get-sdist
         run: |
-          cd dist
-          ls | grep *.tar.gz | xargs -I x mv -- x "$(ls | grep *.tar.gz | sed 's/codecov-cli/codecov_cli/')"
-          ls -al
-          cd ..
+          echo "sdist_filename=$(ls dist/)" >> "${GITHUB_OUTPUT}"
+        shell: bash
+      - name: Build wheels
+        uses: pypa/cibuildwheel@v2.22.0
+        with:
+          package-dir: dist/${{ steps.get-sdist.outputs.sdist_filename }}
+      - name: Store the distribution packages
+        uses: actions/upload-artifact@v4
+        with:
+          name: cibw-wheels-${{ matrix.os }}
+          path: ./wheelhouse/*.whl
+
+
+  publish_to_pypi:
+    if: github.ref_type == 'tag'
+    needs:
+      - build_dist_for_pypi
+      - build_src_for_pypi
+    permissions:
+      id-token: write  # This is required for OIDC
+    runs-on: ubuntu-latest
```

```suggestion
    runs-on: ubuntu-latest
    environment:
      name: pypi
      url: https://pypi.org/p/codecov
```

If you extract version without the leading `v`, you can improve the URL to look smth like `https://pypi.org/project/codecov/<version_here>`

---

@webknjaz out of curiosity, why is this necessary? also, do you mean adding an env var `version` or updating the `url` env var?

---

@thomasrockhu-codecov these *are not* env vars. I think, this confusion is GitHub's fault in terms of naming their features. The `environment:` setting corresponds to Deployment Environments within GitHub (https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-deployments/viewing-deployment-history / https://docs.github.com/en/rest/deployments/deployments?apiVersion=2022-11-28#about-deployments). Env vars are set via `env:`. These are different things and are entirely unrelated to each other.

The environment name will be included in the OIDC payload sent to PyPI. It's best to also set this environment name (`pypi`) in the trusted publisher on PyPI. When set, it'll only allow uploads from jobs that have this environment set, further locking down things that can publish the tokenless way.

The deployment entries will show up @ the corresponding repo tab. Here's an example: https://github.com/cherrypy/cheroot/deployments/pypi. If the `url:` is set, it is also displayed there. The URL is also shown on the workflow run summary page in the graph, e.g., https://github.com/cherrypy/cheroot/actions/runs/8786461123. The URL is more of a convenience thing, it doesn't participate in the checks.

When the workflow will first run this job, it'll auto-create the `pypi` environment in the repo settings. But you can also do this manually. Once created, I recommend setting required reviewers there (https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-deployments/managing-environments-for-deployment#required-reviewers). https://github.com/codecov/codecov-cli/settings/environments. The environments can also have dedicated scoped secrets if you ever need this feature in the future, but it's not something necessary in the context of this PR.

Let me know if you have any ideas on making this all clearer in my guide @ https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/#defining-a-workflow-job-environment.

---

> or updating the `url` env var?

Updating `url` entry, it's not an env var.